File: julia/test/test_provider.jl
# The cadence Provider (esio-9nb.5) — the acceptance: "Provider over the shared
# fixture returns native arrays matching the Python track; CONST/DISCRETE
# correct; refresh_times() matches the cadence." Drives the FULL component
# (a)+(b) pipeline OFFLINE: cache (shared corpus) → format reader → native
# arrays, plus the cadence surface the solver consumes
# (materialize/refresh/refresh_times/prefetch). Reuses the corpus comparison
# helpers from test_readers.jl (included first by runtests.jl).

@testset "cadence Provider — materialize/refresh/refresh_times/prefetch (offline)" begin
    store = LocalStore(joinpath(CORPUS, "cache"))
    cache = Cache(store; offline = true, verify = true)
    era5 = "https://data.earthsci.dev/era5/2018/11/20181108.nc"
    openaq = "https://openaq-data-archive.s3.amazonaws.com/records/openaq/locationid=1/2018-11-08.csv"
    era5_case = JSON.parsefile(joinpath(CORPUS, "cases", "era5-grid-sub-tile.json"))

    @testset "CONST grid: empty cadence, native arrays match the oracle" begin
        p = const_provider(cache, era5; format = "netcdf", source_loader = "era5")
        @test is_const(p)
        @test refresh_times(p) == Float64[]          # CONST ⇒ never refreshes

        nds = materialize(p)
        @test Set(variable_names(nds)) == Set(["t2m", "sp"])
        @test Set(coord_names(nds)) == Set(["latitude", "longitude", "time"])

        # full native-array equality vs the Python track (checks 3–4 via Provider)
        for group in ("variables", "coords")
            for (name, spec) in era5_case["expected"][group]
                @test cmp_native_numeric(nds[name].data, spec["data"]) === nothing
            end
        end
        # the raw time axis is undecoded with its calendar carried for ESS
        @test eltype(nds["time"].data) == Int32
        @test nds["time"].attrs["calendar"] == "gregorian"
    end

    @testset "DISCRETE grid: refresh_times match cadence, per-tick slice" begin
        # cadence taken from the file's own (raw) time axis: [0.0, 1.0]
        full = materialize(const_provider(cache, era5; format = "netcdf"))
        times = Float64.(full["time"].data)
        p = discrete_provider(cache, era5, times; format = "netcdf", time_dim = "time")

        @test !is_const(p)
        @test refresh_times(p) == times              # matches the cadence

        s0 = refresh(p, 0.0)
        s1 = refresh(p, 1.0)
        @test s0["t2m"].dims == ["latitude", "longitude"]   # time record sliced out
        @test size(s0["t2m"].data) == (3, 3)
        @test s0["t2m"].data[1, 1] ≈ 282.5
        @test s1["t2m"].data[1, 1] ≈ 282.6           # a different record per tick
        @test isnan(s1["t2m"].data[3, 3])            # the masked cell survives the slice
        @test !haskey(s0, "time")                    # the sliced dim's coord is dropped
        @test refresh(p, 0.0)["sp"].data == s0["sp"].data   # refresh == materialize

        # a tick between grid points resolves to the active (last ≤ t) record
        @test materialize(p, 0.5)["t2m"].data == s0["t2m"].data
    end

    @testset "DISCRETE per-tick URLs (url-function form, no internal slice)" begin
        # url resolver form: the same fixture stands in for every tick; without
        # time_dim the provider returns the file's full native arrays per tick.
        p = discrete_provider(cache, _ -> era5, [0.0, 1.0]; format = "netcdf")
        @test refresh_times(p) == [0.0, 1.0]
        @test size(refresh(p, 1.0)["t2m"].data) == (2, 3, 3)
    end

    @testset "CSV points provider + variable selection" begin
        p = const_provider(cache, openaq; format = "csv", source_loader = "openaq",
                           reader_kwargs = (numeric_columns = ["latitude", "longitude", "value"],),
                           variables = ["value", "location_id"])
        nds = materialize(p)
        @test Set(variable_names(nds)) == Set(["value", "location_id"])   # restricted
        @test nds["value"].data == [152.3, 168.7, 98.1, 110.4]
        @test eltype(nds["value"].data) == Float64
        @test nds["location_id"].data == ["1", "1", "2", "2"]             # digit text stays string
    end

    @testset "prefetch warms the cache (offline hits, no decode)" begin
        p = const_provider(cache, era5; format = "netcdf")
        entries = prefetch(p)
        @test length(entries) == 1
        @test entries[1].status == :hit

        # DISCRETE per-tick URLs that collapse to one unique blob ⇒ one fetch
        pd = discrete_provider(cache, _ -> era5, [0.0, 1.0]; format = "netcdf")
        @test length(prefetch(pd)) == 1
    end

    @testset "construction + use guards" begin
        @test_throws ArgumentError const_provider(cache, era5; format = "netcdf", times = [1.0])
        @test_throws ArgumentError discrete_provider(cache, era5, Float64[]; format = "netcdf")
        @test_throws ArgumentError const_provider(cache, era5; format = "nonesuch")
        @test_throws ArgumentError Provider(cache, era5; format = "netcdf", time_dim = "time")  # CONST + time_dim
        # a DISCRETE provider needs an explicit time
        @test_throws ArgumentError materialize(discrete_provider(cache, era5, [0.0]; format = "netcdf"))
        # selecting a variable absent from the blob is a clear error
        bad = const_provider(cache, era5; format = "netcdf", variables = ["nope"])
        @test_throws ArgumentError materialize(bad)
    end
end

File: tests/test_provider.py
"""The cadence Provider over the shared corpus — CONST/DISCRETE, OFFLINE.

Acceptance (``esio-9nb.3``): "A Provider over a fixture loader returns correct
native arrays for CONST (materialize once) and DISCRETE (refresh at each anchor);
refresh_times() matches the temporal cadence; unit-tested standalone (no campfire
dependency to pass)." Drives the full (a)+(b) pipeline offline: cache (shared
corpus) → format reader → native arrays, plus the cadence surface the solver
consumes (materialize/refresh/refresh_times/prefetch). Mirrors the peer
``julia/test/test_provider.jl`` and ``rust/tests/provider_cadence.rs``.
"""

from __future__ import annotations

import datetime as dt
import json
import math
import pathlib

import numpy as np
import pytest

from earthsciio import (
    BackendNotRegistered,
    Cache,
    DataLoader,
    LoaderTemporal,
    Provider,
)

REPO_ROOT = pathlib.Path(__file__).resolve().parent.parent
CORPUS = REPO_ROOT / "conformance" / "corpus"

ERA5_URL = "https://data.earthsci.dev/era5/2018/11/20181108.nc"
OPENAQ_URL = (
    "https://openaq-data-archive.s3.amazonaws.com/records/openaq/"
    "locationid=1/2018-11-08.csv"
)

START = dt.datetime(2018, 11, 8, tzinfo=dt.timezone.utc)
HOUR = dt.timedelta(hours=1)
DAY = dt.timedelta(days=1)


def _utc(h: int) -> dt.datetime:
    return START + h * HOUR


@pytest.fixture
def cache():
    """A read-only offline cache rooted at the conformance corpus (no network)."""
    return Cache(root=CORPUS / "cache", offline=True, verify=True)


def _era5_case() -> dict:
    return json.loads((CORPUS / "cases" / "era5-grid-sub-tile.json").read_text())


def _flat(x):
    if isinstance(x, list):
        for e in x:
            yield from _flat(e)
    else:
        yield x


def _match(field, expected_nested) -> bool:
    exp = np.array(
        [math.nan if v is None else float(v) for v in _flat(expected_nested)],
        dtype="f8",
    )
    got = np.asarray(field.data, dtype="f8").reshape(-1)
    if got.shape != exp.shape:
        return False
    gn, en = np.isnan(got), np.isnan(exp)
    return bool(np.array_equal(gn, en) and np.allclose(got[~gn], exp[~en], atol=1e-6, rtol=1e-9))


# --------------------------------------------------------------------------- #
# CONST grid: empty cadence, materialize once, native arrays match the oracle.
# --------------------------------------------------------------------------- #


def test_const_provider_materializes_oracle_arrays(cache):
    case = _era5_case()
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL), cache)
    assert p.is_const
    assert p.refresh_times() == []  # CONST => never refreshes

    nds = p.materialize()
    assert nds.variable_names() == ["sp", "t2m"]
    assert nds.coord_names() == ["latitude", "longitude", "time"]
    # full native-array equality vs the oracle (checks 3–4 through the Provider)
    for group in ("variables", "coords"):
        for name, spec in case["expected"][group].items():
            assert _match(nds[name], spec["data"]), name
    # the raw time axis is undecoded with its calendar carried for ESS
    assert nds["time"].data.dtype == np.int32
    assert nds["time"].attrs["calendar"] == "gregorian"
    # coords property exposes the current grid
    assert set(p.coords) == {"latitude", "longitude", "time"}


def test_const_refresh_returns_constant_data(cache):
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL), cache)
    a = p.refresh(_utc(0))  # CONST: refresh is the constant data (materialize once)
    b = p.refresh(_utc(99))
    assert np.array_equal(a["sp"].data, b["sp"].data)
    assert a["t2m"].shape == (2, 3, 3)


# --------------------------------------------------------------------------- #
# DISCRETE grid: refresh_times match cadence; per-anchor record slice.
# --------------------------------------------------------------------------- #


def _discrete_internal_axis() -> DataLoader:
    # one file holds the day's hourly records; cadence slices the internal axis
    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=2 * HOUR)
    return DataLoader("era5", "netcdf", ERA5_URL, temporal=temporal)


def test_discrete_refresh_times_match_cadence(cache):
    p = Provider(_discrete_internal_axis(), cache, window=(_utc(0), _utc(2)))
    assert not p.is_const
    assert p.refresh_times() == [_utc(0), _utc(1)]  # the hourly cadence


def test_discrete_refresh_slices_record_per_anchor(cache):
    p = Provider(_discrete_internal_axis(), cache, window=(_utc(0), _utc(2)))
    s0 = p.refresh(_utc(0))
    s1 = p.refresh(_utc(1))
    # the time record is sliced out: (time, lat, lon) -> (lat, lon)
    assert list(s0["t2m"].dims) == ["latitude", "longitude"]
    assert s0["t2m"].shape == (3, 3)
    assert s0["t2m"].data[0, 0] == pytest.approx(282.5)
    assert s1["t2m"].data[0, 0] == pytest.approx(282.6)  # a different record per tick
    assert math.isnan(s1["t2m"].data[2, 2])  # the masked cell survives the slice
    assert "time" not in s0  # the sliced dim's coordinate is dropped
    assert set(p.coords) == {"latitude", "longitude"}  # time dropped from coords too


def test_discrete_materialize_primes_first_anchor(cache):
    p = Provider(_discrete_internal_axis(), cache, window=(_utc(0), _utc(2)))
    primed = p.materialize()
    assert np.array_equal(primed["t2m"].data, p.refresh(_utc(0))["t2m"].data)


def test_discrete_between_anchors_uses_active_record(cache):
    p = Provider(_discrete_internal_axis(), cache, window=(_utc(0), _utc(2)))
    at_anchor = p.refresh(_utc(0))["t2m"].data.copy()
    between = p.refresh(START + dt.timedelta(minutes=30))  # snaps down to hour 0
    assert np.array_equal(between["t2m"].data, at_anchor)


def test_discrete_refresh_is_idempotent_within_interval(cache):
    p = Provider(_discrete_internal_axis(), cache, window=(_utc(0), _utc(2)))
    a = p.refresh(_utc(1))["sp"].data
    b = p.refresh(_utc(1))["sp"].data
    assert np.array_equal(a, b)


# --------------------------------------------------------------------------- #
# refresh_times bounds: window end, temporal.end, and the unbounded case.
# --------------------------------------------------------------------------- #


def test_refresh_times_bounded_by_temporal_end_without_window(cache):
    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=2 * HOUR, end=_utc(3))
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL, temporal=temporal), cache)
    assert p.refresh_times() == [_utc(0), _utc(1), _utc(2)]


def test_refresh_times_empty_when_unbounded(cache):
    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=2 * HOUR)
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL, temporal=temporal), cache)
    assert p.refresh_times() == []  # no window, no end => no enumerable schedule


def test_refresh_times_window_start_clamped_to_epoch(cache):
    # window starts mid-cadence; the first tstop is the aligned anchor >= start
    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=DAY)
    p = Provider(
        DataLoader("era5", "netcdf", ERA5_URL, temporal=temporal),
        cache,
        window=(START + dt.timedelta(minutes=30), _utc(3)),
    )
    assert p.refresh_times() == [_utc(1), _utc(2)]


# --------------------------------------------------------------------------- #
# Per-file URL resolution (strftime template + callable) and prefetch.
# --------------------------------------------------------------------------- #


def test_strftime_url_template_resolves_per_anchor():
    loader = DataLoader("era5", "netcdf", "https://data.earthsci.dev/era5/%Y/%m/%Y%m%d.nc")
    assert loader.resolve_url(START) == ERA5_URL


def test_prefetch_const_warms_single_file(cache):
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL), cache)
    entries = p.prefetch()
    assert len(entries) == 1
    assert entries[0].status == "hit"  # offline corpus hit, no decode


def test_prefetch_strftime_window_hits_corpus(cache):
    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=DAY)
    loader = DataLoader("era5", "netcdf", "https://data.earthsci.dev/era5/%Y/%m/%Y%m%d.nc",
                        temporal=temporal)
    p = Provider(loader, cache, window=(START, START + DAY))
    entries = p.prefetch()
    assert len(entries) == 1 and entries[0].status == "hit"


def test_prefetch_enumerates_file_anchors_and_dedups(cache):
    seen = []

    def url_for(anchor):
        seen.append(anchor)
        return ERA5_URL  # every file anchor collapses to the one corpus blob

    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=HOUR)
    p = Provider(DataLoader("era5", "netcdf", url_for, temporal=temporal), cache,
                 window=(_utc(0), _utc(2)))
    entries = p.prefetch()
    assert seen == [_utc(0), _utc(1)]  # one anchor per file period across the window
    assert len(entries) == 1  # collapsed to a single unique fetch
    assert entries[0].status == "hit"


def test_prefetch_unbounded_raises(cache):
    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=HOUR)
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL, temporal=temporal), cache)
    with pytest.raises(ValueError):
        p.prefetch()  # no window, no temporal.end


# --------------------------------------------------------------------------- #
# CSV points provider + variable selection (the 2nd format, end to end).
# --------------------------------------------------------------------------- #


def test_csv_provider_with_variable_selection(cache):
    loader = DataLoader(
        "openaq",
        "csv",
        OPENAQ_URL,
        variables=["value", "location_id"],
        reader_kwargs={"numeric_columns": ["latitude", "longitude", "value"]},
    )
    nds = Provider(loader, cache).materialize()
    assert nds.variable_names() == ["location_id", "value"]  # restricted
    assert list(nds["value"].data) == [152.3, 168.7, 98.1, 110.4]
    assert nds["value"].data.dtype == np.float64
    assert nds["location_id"].data == ["1", "1", "2", "2"]  # digit text stays string


# --------------------------------------------------------------------------- #
# Construction + use guards.
# --------------------------------------------------------------------------- #


def test_unknown_format_raises_at_construction(cache):
    with pytest.raises(BackendNotRegistered):
        Provider(DataLoader("x", "nonesuch", ERA5_URL), cache)


def test_loader_temporal_rejects_nonpositive_cadence():
    with pytest.raises(ValueError):
        LoaderTemporal(start=START, frequency=dt.timedelta(0), file_period=HOUR)
    with pytest.raises(ValueError):
        LoaderTemporal(start=START, frequency=HOUR, file_period=dt.timedelta(0))


def test_refresh_before_start_raises(cache):
    p = Provider(_discrete_internal_axis(), cache, window=(_utc(0), _utc(2)))
    with pytest.raises(ValueError):
        p.refresh(START - HOUR)


def test_refresh_record_out_of_range_raises(cache):
    # file_period=DAY claims 24 hourly records, but the fixture file holds 2
    temporal = LoaderTemporal(start=START, frequency=HOUR, file_period=DAY)
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL, temporal=temporal), cache)
    with pytest.raises(IndexError):
        p.refresh(_utc(5))  # record 5 absent from the 2-record file


def test_absent_variable_raises(cache):
    p = Provider(DataLoader("era5", "netcdf", ERA5_URL, variables=["nope"]), cache)
    with pytest.raises(KeyError):
        p.materialize()
